File: sources/CPP/include/Ers/Model/ModelManager.cpp
#include "Ers/Model/ModelManager.h"
#include "Ers/Api.h"
#include "Ers/Logger.h"
#include "Ers/Model/ModelContainer.h"

#include <cassert>
#include <functional>
#include <stdexcept>

namespace Ers
{
    /// @brief Add the given ModelContainer to this ModelManager.
    /// The ModelManager will update the ModelContainer until it reaches the ReleaseTime
    /// @param modelContainer The given ModelContainer to add.
    /// @param releaseTime When this time is reached the ModelContainer is removed from the ModelManager.
    void ModelManager::AddModelContainer(ModelContainer& modelContainer, SimulationTime releaseTime)
    {
        assert(modelContainer.Valid());
        assert(!HasModelContainer(modelContainer));
        Ers::Engine::ERS_ModelManager_AddModelContainer(this, modelContainer.CorePtr(), releaseTime);
    }

    /// @brief Remove the given ModelContainer from this ModelManager.
    /// NOTE: This will invalidate of the given ModelContainer.
    /// @param modelContainer The given ModelContainer to remove.
    void ModelManager::RemoveModelContainer(ModelContainer& modelContainer)
    {
        assert(modelContainer.Valid());
        if (!HasModelContainer(modelContainer))
        {
            Ers::Logger::Error("ModelContainer is not inside this ModelContainer");
            return;
        }

        Ers::Engine::ERS_ModelManager_RemoveModelContainer(this, modelContainer.CorePtr());
    }

    size_t ModelManager::Count()
    {
        return Ers::Engine::ERS_ModelManager_CountModelContainers(this);
    }

    /// @brief Returns a shared pointer to the ModelContainer at the given position in the vector of ModelContainers.
    /// @param index The index of the ModelContainer to get.
    /// @return Returns a shared pointer to the ModelContainer at the given position.
    /// @throws If no ModelContainer is found at given index, e.g. if index is out of bound.
    ModelContainer ModelManager::GetModelContainerAt(const std::size_t& index)
    {
        void* coreModelContainerPtr = Ers::Engine::ERS_ModelManager_GetModelContainerAt(this, index);
        return ModelContainer(coreModelContainerPtr);
    }

    /// @brief Checks whether the given ModelContainer is present in the collection of the ModelContainers of this ModelManager.
    /// @param modelContainer The given ModelContainer to check.
    /// @return Returns true if the given ModelContainer is present in the vector of ModelContainers associated with this ModelManager,
    /// otherwise returns false
    bool ModelManager::HasModelContainer(const ModelContainer& modelContainer) const
    {
        return Ers::Engine::ERS_ModelManager_HasModelContainer(this, modelContainer.CorePtr());
    }

    //============== Protected functions ====================

    /// @brief Performs Update for the ModelContainers of this ModelManager by invoking the corresponding function of the core.
    void ModelManager::Update()
    {
        Ers::Engine::ERS_ModelManager_Update(this);
    }

    void ModelManager::RunWithProgressBar()
    {
        Ers::Engine::ERS_RunWithProgressBar();
    }

    ModelManager& ModelManager::Get()
    {
        return *static_cast<ModelManager*>(Ers::Engine::ERS_ModelManager_GetModelManager());
    }

} // namespace Ers

File: sources/CSharp/src/Ers/Model/ModelManager.cs
using System.Diagnostics;
using System.Runtime.InteropServices;
using Ers.Engine;

namespace Ers
{
    /// <summary>
    /// Class overseeing simulation models (<see cref="ModelContainer"/>.
    /// It will manage and update the active collection of ModelContainers.
    /// </summary>
    [StructLayout(LayoutKind.Sequential)]
    public ref struct ModelManager
    {
        /// <summary>
        /// Native pointer to the core instance.
        /// </summary>
        public readonly IntPtr CorePtr;

        /// <summary>
        /// Construct a new ModelManager.
        ///
        /// <para>Use <see cref="Get"/> instead of this constructor.</para>
        /// </summary>
        public ModelManager() { CorePtr = ErsEngine.ERS_ModelManager_GetModelManager(); }

        /// <summary>
        /// Get the ModelManager singleton.
        /// </summary>
        /// <returns></returns>
        public static ModelManager Get() => new ModelManager();

        /// <summary>
        /// Add a <see cref="ModelContainer"/> to the manager.
        /// </summary>
        /// <param name="modelContainer">The ModelContainer to add.</param>
        /// <param name="endTime">The end time for the ModelContainer.</param>
        public void AddModelContainer(in ModelContainer modelContainer, SimulationTime endTime)
        {
            Debug.Assert(!HasModelContainer(modelContainer));
            ErsEngine.ERS_ModelManager_AddModelContainer(CorePtr, modelContainer.CorePtr, endTime);
        }

        /// <summary>
        /// Remove a ModelContainer from the manager.
        /// </summary>
        /// <param name="modelContainer"></param>
        /// <exception cref="Exception">When the given ModelContainer is not in the manager.</exception>
        public void RemoveModelContainer(in ModelContainer modelContainer)
        {
            if (!HasModelContainer(modelContainer))
            {
                throw new Exception("ModelContainer is not inside this ModelContainer");
            }

            ErsEngine.ERS_ModelManager_RemoveModelContainer(CorePtr, modelContainer.CorePtr);
        }

        /// <summary>
        /// Check whether a given ModelContainer is in the manager.
        /// </summary>
        /// <param name="modelContainer">The ModelContainer to check for.</param>
        /// <returns>True if the ModelContainer exists in the manager, false if not.</returns>
        public bool HasModelContainer(in ModelContainer modelContainer)
        {
            return ErsEngine.ERS_ModelManager_HasModelContainer(CorePtr, modelContainer.CorePtr);
        }

        /// <summary>
        /// Get the model container at the given index.
        /// </summary>
        /// <param name="index"></param>
        /// <returns></returns>
        public ModelContainer GetModelContainerAt(nuint index)
        {
            IntPtr coreModelContainerPtr = ErsEngine.ERS_ModelManager_GetModelContainerAt(CorePtr, index);
            return new ModelContainer(coreModelContainerPtr);
        }

        /// <summary>
        /// The number of <see cref="ModelContainer"/>s in this ModelManager.
        /// </summary>
        /// <returns></returns>
        public ulong Count
        {
            get => ErsEngine.ERS_ModelManager_CountModelContainers(CorePtr);
        }

        /// <summary>
        /// Update the manager, updating all ModelContainers inside it.
        /// </summary>
        public void Update() => ErsEngine.ERS_ModelManager_Update(CorePtr);

        /// <summary>
        /// Run all models that are currently in the ModelManager until completion and print a combined progress bar while running.
        /// </summary>
        public void RunWithProgressBar() => ErsEngine.ERS_RunWithProgressBar();
    }
}
